File: Homework/Ngan_Luu/148-sort-list.py
https://leetcode.com/problems/sort-list/ (implement merge sort)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self,head):
        slow = head
        fast = head.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow
        
    def merge(self,left,right):
        sorted_list=temp=ListNode()
        while left and right:
            if left.val < right.val:
                sorted_list.next=left
                left=left.next
            else:
                sorted_list.next=right
                right=right.next
            sorted_list=sorted_list.next
        if left:
            sorted_list.next=left
        if right:
            sorted_list.next=right
        return temp.next

    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        left=head
        middle_node=self.middleNode(head)
        right=middle_node.next
        middle_node.next=None
        
        left=self.sortList(left)
        right=self.sortList(right)
        return self.merge(left,right)

File: Homework/Duy_Truong/Lesson3/148. Sort List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function(head) {
    if(!head || !head.next) return head;
    
    let fast=head.next.next,slow=head;
    
    while(fast && fast.next){
        fast=fast.next.next;
        slow=slow.next;
    }
    
    const mid=slow.next;
    slow.next=null;
    
    let left = sortList(head);
    let right = sortList(mid);
    
    return merge(left,right);
};

function merge(left,right){
	let dummy = new ListNode();
	const headPtr = dummy;
	while(left && right){
		if(left.val < right.val){
			dummy.next = left;
			left = left.next
		} else{
			dummy.next = right;
			right = right.next;
		}
		dummy = dummy.next;
	}
	dummy.next = (right) ? right : left
	return headPtr.next;

}